src/nccl_ofi_rdma.c

```diff
@@ -4821,8 +4832,9 @@ static int blocked_send_close(nccl_net_ofi_send_comm_t *send_comm)
 		return ncclInternalError;
 	}
 
-	// TODO: We might want to use READ_ONCE to read variable `connected'
-	while (!s_comm->connected) {
+	// TODO: We might want to use READ_ONCE to read variables
+	// `connect_msg_sent` and `connected'
+	while (!s_comm->connect_msg_sent || !s_comm->connected) {
```

Comment:
this is a weird set of flags now.  Are we sure this is the right progression of state machine?

Reply:
I was just mirroring how we handle `s_comm->connected`. This line is preventing closing the communicator while any connection establishment operations are still in progress.

Reply:
This is really a walk through a state machine; having a bunch of different flags for where you are in the state machine is kind of weird.

Reply:
I agree, the new flags threw me off too. Reading this code for the first time, looks like you want a way to peek into the state machine from outside the connection management calls - progress or send etc. You can just carry the state in the s_comm itself and check to see if it is `COMM_CONNECTED` to decide if you want to progress connection requests before scheduling sends.

Reply:
I think I understand. I think the confusing bit is, with this PR we post the connection response recv without waiting for the connect message to be delivered, which really leads to two independent events (connect message delivered and connect response received) that could happen in either order, rather than a state machine where the first must happen before the second.

Reply:
Then "connected" is a bit of a misnomer in this protocol.  If you really do have two states that can happen in random order, flags are ok, but the names have to be meaningful.

Reply:
Renamed `connect_resp_finalized`